### PR TITLE
bpo-38530: Clean exceptions if dir() fails when making suggestions for AttributeError

### PR DESCRIPTION
**Pull request by pablogsal**
Caught by the mighty eye of @dutc!

https://bugs.python.org/issue38530

### CODE REVIEW

File: Lib/test/test_exceptions.py [hunk 1]
```diff
@@ -1695,6 +1695,19 @@ def __getattr__(self, attr):
 
             self.assertIn("blech", err.getvalue())
 
+    def test_attribute_error_with_failing_dict(self):
+        class T:
+            bluch = 1
+            def __dir__(self):
+                return self.blech
```
Comment (vstinner) [inline Lib/test/test_exceptions.py]: I would prefer to explicitly raise an exception. It's hard to guess the intent here.

**Review by vstinner — COMMENTED** [review]
Honestly, I would prefer to clear the exception in the function calling _Py_Offer_Suggestions().

**Comment by pablogsal** [comment]
> Honestly, I would prefer to clear the exception in the function calling _Py_Offer_Suggestions().

Ok, we can do that

File: Lib/test/test_exceptions.py (hunk 1 above)
Comment (pablogsal) [reply Lib/test/test_exceptions.py]: Done

File: Lib/test/test_exceptions.py [hunk 2]
```diff
@@ -1695,6 +1695,19 @@ def __getattr__(self, attr):
...
+            def __dir__(self):
+                raise AttributeError("oh no!")
+
+        try:
+            T().blich
+        except AttributeError as exc:
+            with support.captured_stderr() as err:
+                sys.__excepthook__(*sys.exc_info())
+
+        self.assertNotIn("blech", err.getvalue())
```
Comment (ammaraskar) [inline Lib/test/test_exceptions.py:1710]: This should be `"oh no!"` now, right?

Comment (pablogsal) [reply Lib/test/test_exceptions.py:1710]: The "blech" is the name of the attribute for the suggestion, but we can test the exception as well.

Comment (pablogsal) [reply Lib/test/test_exceptions.py:1710]: Done in 34fbacb

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**